include/psa/crypto_values.h

```diff
@@ -2021,6 +2021,15 @@
 #define PSA_ALG_TLS12_PSK_TO_MS_GET_HASH(hkdf_alg)                         \
     (PSA_ALG_CATEGORY_HASH | ((hkdf_alg) & PSA_ALG_HASH_MASK))
 
+/* Macro to build a KDF that takes the shared secret K (an EC point in case
+ * of EC J-PAKE) and calculates SHA256(K.X) that the rest of TLS 1.2 will
+ * use to derive the session secret. Uses PSA_ALG_SHA_256. Only P-256 is
+ * supported, so the input has to be exactly 65 bytes.
+ */
```

Comment:
Perhaps also add this KDF only supports output in a single chunk of 32 bytes. @gilles-peskine-arm is that an acceptable restriction? (The capacity is limited to 32 bytes, that's not a question, but what I mean is we don't want to support reading it in two chunks of 16 bytes, or 1+31 or whatever: you're supposed to request 32 bytes at once, period.)

Reply:
Not added yet, as it's a new discussion.

Reply:
I started this discussion because it looks to me that's what the code does - see [this comment](https://github.com/Mbed-TLS/mbedtls/pull/6115#discussion_r979593467) from @mprse who noticed the same thing.

Whatever we decide, the documentation and code need to be in agreement: either the code has this limitation and the documentation mentions it, or the documentation doesn't mention it and the code needs to support 31+1 output or even 1+1+...+1.

Since that would require more code, to keep track of where we are in the output stream, and also more tests, and we don't need that for the only use case this function was written for in the first place, I'd say let's go for the simplest thing that works and restrict the output to "32 bytes in one go and that's it". (Also, from a backwards compat perspective, it we're strict now we can go more lenient later, but not the opposite.)

Reply:
Restricting the output to a single chunk makes sense here. But then we could also have done that for the other TLS-oriented KDFs and didn't. So the code to handle a KDF with fixed output and split it is already present.

[I'll think about improving the driver interface to handle this more seamlessly](https://github.com/Mbed-TLS/mbedtls/issues/6335).

Reply:
The `psa_key_derivation_output_key()` function already has a wide range of examples that permit an implementation to report `INVALID_ARGUMENT` - which can include an invalid key size used for the output.

The same error code is not listed against `psa_key_derivation_output_bytes()`, which indicates an expectation that clients can extract the stream in arbitrary-sized blocks.

However, I am not entirely against the idea that some KDFs do not support, or permit, incremental output. This would be particularly relevant for KDFs that have fixed output sizes (such as a this one).

Would a restriction on only permitting key output be viable for KDFs such as this one?

Reply:
> Would a restriction on only permitting key output be viable for KDFs such as this one?

I don't think that would make sense. It's natural to implement `output_key` on top of `output_bytes`, and that's official for accelerator drivers which don't have an `output_key` entry point.

What could make sense is to restrict some KDFs to be one-shot in the sense that you may only call one `output_xxx` function, once. But I'm not sure it's the right way to go. It isn't that uncommon to use a KDF to generate e.g. a key and an IV successively. I don't know how many other KDFs we'll add that are similar to the TLS-specific KDFs. Or who might have repurposed the TLS-specific KDFs in another protocol. I wish KDFs weren't such a mess…

For a KDF with small, fixed-size output, it isn't a big deal to keep the derived data in memory and have code to serve it piece by piece. The relevant code is not very big and is likely to exist in some form for longer-output KDFs such as HKDF anyway.

Reply:
> The same error code is not listed against `psa_key_derivation_output_bytes()`, which indicates an expectation that clients can extract the stream in arbitrary-sized blocks.

I think that's a pretty strong argument, thanks for pointing this out. Let's allow piecewise output then.